src/java/org/apache/cassandra/dht/RangeStreamer.java

```diff
@@ -265,10 +318,11 @@ public void addRanges(String keyspaceName, ReplicaCollection<?> replicas)
             workMap = getOptimizedWorkMap(fetchMap, sourceFilters, keyspaceName);
         }
 
-        toFetch.put(keyspaceName, workMap);
-        for (Map.Entry<InetAddressAndPort, Collection<FetchReplica>> entry : workMap.asMap().entrySet())
+        assert toFetch.put(keyspaceName, workMap) == null : "Keyspace is already added to fetch map";
```

Comment:
If asserts are disabled this doesn't work :-P

Reply:
Fixed